docs/develop/web-client-performance.md

```diff
@@ -0,0 +1,107 @@
+---
+title: 'Improve the performance of your Office Scripts'
+description: 'How to create faster scripts through understanding how the script communicates with the Excel workbook.'
+ms.date: 05/14/2020
+localization_priority: Normal
+---
+
+# Improve the performance of your Office Scripts
+
+The promise of Office Scripts is to automate commonly performed series of tasks to save time. A slow script can feel like it invalidates this promise. Most of the time, your script will be perfectly fine and run as expected. However, there are a couple, avoidable scenarios that can affect performance.
+
+The most common reason for a slow script is excessive communication with the workbook. Your script runs on your local machine, while the workbook exists in the cloud. At certain times, your script synchronizes its local data with that of the workbook. This means that any write operations (such as `workbook.addWorksheet()`) are only applied to the workbook when this behind-the-scenes synchronization happens. Likewise, any read operations (such as `myRange.getValues()`) only get data from the workbook for the script at those times. In either case, the script fetches information before it acts on the data. For example, the following code will accurately log the number of rows in the used range.
+
+```TypeScript
+let usedRange = workbook.getActiveWorksheet().getUsedRange();
+let rowCount = usedRange.getRowCount();
+// The script will read the range and row count from
+// the workbook before logging the information.
+console.log(rowCount);
+```
+
+Office Scripts APIs ensure any data in the workbook or script is accurate and up-to-date when necessary. You don't need to worry about these synchronizations for your script to run correctly. However, an awareness of this script-to-cloud communication can help you avoid unneeded network calls.
+
+## Performance optimizations
+
+You can apply simple techniques to help reduce the communication to the cloud. The following patterns help speed up your scripts.
+
+- Read workbook data once instead of repeatedly in a loop.
+- Remove unnecessary `console.log` statements.
+- Avoid using try/catch blocks.
+
+### Read workbook data outside of a loop
+
+Any method that gets data from the workbook can trigger a network call. Rather than repeatedly making the same call, you should save data locally whenever possible. This is especially true when dealing with loops.
+
+Consider a script to get the count of negative numbers in the used range of a worksheet. The script needs to iterate over every cell in the used range. To do that, it needs the range, the number of rows, and the number of columns. You should store those as local variables before starting the loop. Otherwise, each iteration of the loop will force a return to the workbook.
```

Comment:
Good example.
Is should too proscriptive? Would it be better to say something like "You can speed up you script by..."

Reply:
We want to be prescriptive here. The user is reading the article to improve their script's performance.